### ISSUE TITLE
weekly useful materials -05/18-

### ISSUE THREAD

**Issue by GENZITSU** [GENZITSU/UsefulMaterials#52]

**Comment by GENZITSU** [comment]
## テレビ局のAI活用 ～Interop Tokyo 2021～「テレビ局におけるAI活用事例」レポート

### 日本テレビ
> ロードレース中継での選手の距離測定 + ユニフォームによるチーム識別

選手の移動距離などの把握は以前から要望としてあったらしい。
<img width="600" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/118610031-53c21c00-b7f6-11eb-86d2-6748e9afd9d8.jpg">

顔写真→CG生成を高速に行うことにも活用
<img width="600" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/118610091-62a8ce80-b7f6-11eb-9b04-785beb99610a.jpg">

AIキャッチャー + 得点率予測
<img width="600" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/118610456-be735780-b7f6-11eb-8fd0-d2e7585a4c59.jpg">

OCRからの字幕翻訳
<img width="600" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/118611054-64bf5d00-b7f7-11eb-9fcb-14452bef414e.jpg">

提供元テロップとの文字かぶり防止  
どうも1秒でも文字被りが発生すると放送事故扱いになるとのこと
<img width="600" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/118611133-7b65b400-b7f7-11eb-8bff-b44f58fbe249.jpg">

その他にも、大量のアーカイブ素材から特定人物の登場シーンを検索するシステムやドラマと連動したチャットボット、字幕制作、視聴率の予測にもAIを活用している。

### TBS

文字起こし
<img width="300" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/118999932-a1d94a00-b9c5-11eb-9156-f7235367fc30.jpeg"><img width="300" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/118999946-a43ba400-b9c5-11eb-9f9d-a5f46a992331.jpeg">

そのほかにも
- 権利処理が必要なものの映り込みチェック
- 出演者が映っているシーンの検索
- 動画/静止画からの人物名推定 (約16,000枚)


### 感想
技術進歩によりやる気とデータさえあればそれなりの案件ができてしまうようになった。  
AIの専門家としての優位性はデータ(?) → ビジネス目標のギャップを埋めることのうち。  
泥臭い部分になる気もする...


[出典リンク](https://www.screens-lab.jp/article/26714)

**Comment by GENZITSU** [comment]
## Table data用のTransformer

2020年12月にひっそりとawsから出ていたtableデータ向けtransformerの論文とgithub実装。

この論文では数多のtableデータ向けNNと違って、LGBMに僅かながらも辛勝しているのが特筆するべきところ。

<img width="400" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119219546-1aa0e900-bb21-11eb-9923-5bfd20245a1a.png">


利点としては
- MLPと異なり、似たようなカラムは潜在空間上でも同じような場所にembeddingされるところ
- MLM的なunsupervised learningでpretrainできるところ

pretrainの方法は2種類ある
- MLM: ランダムに特徴をマスクして、復元するタスク 
- RTD: ランダムに特徴を入れ替えて、どの特徴が入れ替わったものかを当てるタスク。ただし、ランダムに入れ替えたサンプルは容易に見破れるので、[元ネタのELECRA](https://openreview.net/forum?id=r1xMH1BtvB)では、encoderを使って候補を生成する。

## 実験結果など

精度面ではMLPよりも1%程度良さげ。
<img width="400" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119219956-f47c4880-bb22-11eb-94bc-c0e768aaf001.png">

解釈性の面ではMLPよりもかなりよい。(Attentionを通すことで、似た意味のcolumnが同じようなところに配置される。)
<img width="600" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119219952-f219ee80-bb22-11eb-9e5c-b83d8b8b354e.png">

labelなしデータが大量にあるケースだと、LGBM系よりも精度が良くなる傾向。  
普通の教師ありの設定だと、同じくらい(多分負けてるくらい)
<img width="400" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119220107-c3e8de80-bb23-11eb-8fb8-cbddbab3b166.png">

## 補足
pretrainingアプローチは、ラベルなしデータが大量にある場合にしか役立たなかったとのとこ。
>We do not find much benefit in using it when the entire data is labeled. Its benefit is evident when there is a large number of unlabeled examples and a few labeled examples. Since in this scenario the pre-training provides a representation of the data that could not have been learned based only on the labeled examples.

[出典論文](https://blog.ml6.eu/transformers-for-tabular-data-hot-or-not-e3000df3ed46)
[githubリンク](https://github.com/lucidrains/tab-transformer-pytorch)

**Comment by GENZITSU** [comment]
## styleganを使って無限のデータセットを得るDataset GAN (CVPR2021)

この図が手法の全てを表している。

StyleGanの各AdaIN層には高次元のなんらかの意味を保持しているのでは？  
ということで、各層の特徴量をupsampleして特徴量マップを画像っぽく整形し、  
少量(10 ~ 20 ~ 40オーダー)のデータで学習したMLPをかませることで、無限にアノテーションデータを生成する。。。
<img width="800" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119221193-1973ba00-bb29-11eb-90a7-6c985ee9fa56.png">

StyleGANはどうもゴミデータを生成してしまうことがあるようで、フィルタリングが必要。   
StyleGANの識別機はあんまり役に立たないが、MLPの出力でJS距離つかって、ゴミデータを識別するとのこと
> Naturally, StyleGAN also fails occasionally which introduces noise in the synthesized dataset. We noticed that the
StyleGAN’s discriminator score is not a robust measure of failure and we found that utilizing our ensemble of classifiers to measure the uncertainty of a synthesized example is a more robust approach. We follow [30], and use the
Jensen-Shannon (JS) divergence [2, 39] as the uncertainty measure for a pixel. To calculate image uncertainty, we sum
over all image pixels. We filter out the top 10% most uncertain images.


## annotationの比較
134時間 vs 5時間(提案手法) !?

>Crowdsourcing labels on the same scale (10k images) for one dataset would take over 3200 hours (134 days), and,
we hypothesize, would be extremely noisy since annotating an image to that level of detail requires both skill and immense patience.

> In our case, human-annotation time for a dataset was roughly 5 hours, affording us to leverage a single skilled annotator. This is described next.


## 生成した結果

<img width="800" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119221481-981d2700-bb2a-11eb-9a7b-bcf317c9bbea.png">

特徴点の方
<img width="800" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119221731-a0299680-bb2b-11eb-9e81-e3bf88adcc62.png">

## 感想

詳細な部位ごとのsegmentationベースだけど、普通に猫/車的なsegmentationもできるんだろうか？

[出典リンク](https://panhouse.blog/paper/introduction/datasetganefficient_labeled_data_factory_with_minimal_human_effort/)

**Comment by GENZITSU** [comment]
## 単語分散表現を扱う上でmagnitudeが有用らしい

なんかmagnitudeなるものを使うと、word2vecの読み込みがめちゃくちゃ早くなるらしい。

<img width="800" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119325684-178e3000-bcbc-11eb-9214-17d543d911ec.png">

[元ネタ](https://twitter.com/goto_yuta_/status/1396700569897627648?s=20)

ということで、magnitudeを少し調べてみた。

### magnitudeって何よ

元々は単語ベクトルのOOV対策のためのツールらしい。  
単語ベクトルのOOV対策としてランダムな割り当てを行うのだが、文字列的に似ている単語動詞のベクトル表現を近づけるような付与を行う。

>magnitudeというPythonパッケージにおいて特徴的なOOV対策がされているのを見つけました。基本的にはOOVにはランダムなベクトル表現を付与するのですが、下記の2つの工夫が入っています。

>文字列的に似ているOOVの単語同士には、なるべく似たベクトル表現を付与したい
e.g. uberの車種であるuberxとuberxlは、似た文字列であるから似たランダムベクトルになってほしい (OOVだけど意味的に似ていると想定される)

>OOVと文字列的に似ている単語が語彙中の中にあれば、そのベクトル表現と似たベクトル表現を付与したい
e.g. uberxというOOVは、uberが分散表現の語彙の中にあれば、それに近いランダムベクトル表現であってほしい (単語を構成する部分文字列の意味と似ていると想定される)

アルゴリズム的には、未知語の単語ベクトルを、未知語の3gram-6gramから作成されるランダムベクトルと文字列的に近い既存単語にtop3のベクトルを用いて作成するみたい。

<img width="400" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119327469-fa5a6100-bcbd-11eb-87d0-f0f8611a4611.png">

<img width="400" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119327474-faf2f780-bcbd-11eb-8add-8c9824829165.png">

[解説記事](https://user-images.githubusercontent.com/19774756/119325684-178e3000-bcbc-11eb-9214-17d543d911ec.png)

### デジタル広告の出稿を決める上でのセグメンテーションにも活用されている。
fasttextとmagnitiudeの比較  
magnitudeを用いることでOOVでありながらカメラに近いベクトルを取得できている。

<img width="300" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119328015-98e6c200-bcbe-11eb-848b-8c57d3035f28.png"><img width="300" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119328019-9a17ef00-bcbe-11eb-909e-deecc61a6f07.png">

[解説記事](https://note.com/dd_techblog/n/ndf2f3a786643)

### 使い方
学習済みの単語ベクトルにmagnitude converterをかければ良いっぽい
`python -m pymagnitude.converter -i <PATH TO FILE TO BE CONVERTED> -o <OUTPUT PATH FOR MAGNITUDE FILE>
`

どうも日本語の計算済みmagnitudeベクターもあるみたい(*[参考記事](https://zenn.dev/sorami/articles/fb2eb78e250568b767fd))

### 感想
magnitude自体は計算済みのベクトルへのOOV対応が主目的っぽい。副産物として読み込み速度が爆速になる模様。  
読み込み速度が早くなるってことはメモリーとの兼ね合いとかどうなるんだろうか...

**Comment by GENZITSU** [comment]
## 教師なしで音声認識を行う手法

非常にざっくりなことをいうと、音声データをベクトル化して音素に分ける。  
リアルなテキストデータから生成される音素と音声から生成される音素を見分けるGANを用いて、生成器(つまり音声認識器)を訓練する。

<img width="700" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119353778-474e2f80-bcde-11eb-87cf-660a988b9500.png">

それぞれのstepに血の滲むような努力がありそう。

### 1. audioデータのベクトル化

Convolutional layerにaudioデータを噛ませた後に24層transformerを通したベクトル表現を用いる。  
事前学習としてはmaskされた箇所のベクトル表現を当てるcontrastive learning。

<img width="600" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119457159-5680bc00-bd76-11eb-8f7e-3fb750205e80.png">

実験ではここのモデルはpretrainedモデルを使ったとのこと
>we use the publicly available English model pre-trained on 53k hours of Libri-Light (Kahn et al., 2020b) as well as XLSR-53 which was pre-trained on nearly 60k hours of speech audio in 53 languages (Conneau et al., 2020). Both use Large transformers with 24 blocks, model dimension 1,024

どうやら、24層あるそれぞれの層のうち、最初の方と最後の方は、speech recognitionに有用な表現ではないらしい。
教師あり学習で確かめたところ、英語以外の言語も含めて15層目を使うのがいいらしい。
<img width="600" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119457628-d870e500-bd76-11eb-89b9-ce620c5031d2.png">

### audioデータのsegmentation

上述のencoderでベクトル化しものをK(=128)-meansクラスタリングを行って、segmentに分ける。  
データ量がやばいのでFAISS+GPUを使ったとのこと。

流石にK-meansで精度が高いものはできないので、以下のことを実施。
1. audio全体でPCAを学習。
2. 対象segment+隣接segmentにPCAをかけて、平均化して、そのsegmentのベクトルとみなす。

これを行うことで、そこそこの音素列になるっぽいが、彼らはこの後viterbiアルゴリズムを適用して予測行っている。
(詳細は正直不明)

<img width="600" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119459735-ecb5e180-bd78-11eb-9a09-99f742d02eb4.png">
<img width="600" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119459756-f0e1ff00-bd78-11eb-99f1-2f3e783f09e3.png">

### テキストデータの音素化

何らかの音素化機を用いて音素化を行う。TIMITデータに関しては元々音素化がされているらしい。(しかも沈黙トークンも付与されている。)  
その他のデータについては、G2P phonemizer (Park and Kim, 2019)なるものを使って音素化するとのこと。

ただし、音声データには沈黙している期間があるので、最初と最後に沈黙記号を付与しつつ、25%の確率でランダムに沈黙記号を入れるなどの工夫が必要。

<img width="600" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/119460511-bb89e100-bd79-11eb-8c34-6dc553931ffb.png">


### 教師なし音声認識
unlabelテキストデータ由来のものか, 音声データ由来のものかを見分けるGAN

![スクリーンショット 2021-05-25 16 55 15](https://user-images.githubusercontent.com/19774756/119460783-fee44f80-bd79-11eb-936c-1bbcbc456f52.png)


### 結果

![スクリーンショット 2021-05-25 16 57 26](https://user-images.githubusercontent.com/19774756/119461070-4ec31680-bd7a-11eb-842c-4fd87bab3cd8.png)





[出典リンク](https://scontent-nrt1-1.xx.fbcdn.net/v/t39.8562-6/187874612_311717527241594_5668815448923437055_n.pdf?_nc_cat=102&ccb=1-3&_nc_sid=ae5e01&_nc_ohc=qXOFpESwsyUAX8Rmzf7&_nc_ht=scontent-nrt1-1.xx&oh=240447556d9c2e98d6b388298ec07b39&oe=60D1D515)
